// unplugin-vue-define-options/dist/chunk-35IND6RP.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkTTU2WSZJjs = require('./chunk-TTU2WSZJ.js');

// src/index.ts
var _unplugin = require('unplugin');
var _pluginutils = require('@rollup/pluginutils');
var _common = require('@vue-macros/common');
function resolveOption(options) {
  return {
    include: options.include || [_common.REGEX_VUE_SFC, _common.REGEX_SETUP_SFC],
    exclude: options.exclude || void 0
  };
}
var src_default = _unplugin.createUnplugin.call(void 0, (userOptions = {}) => {
  const options = resolveOption(userOptions);
  const filter = _pluginutils.createFilter.call(void 0, options.include, options.exclude);
  const name = "unplugin-vue-define-options";
  return {
    name,
    enforce: "pre",
    transformInclude(id) {
      return filter(id);
    },
    transform(code, id) {
      try {
        return _chunkTTU2WSZJjs.transform.call(void 0, code, id);
      } catch (err) {
        this.error(`${name} ${err}`);
      }
    }
  };
});



exports.src_default = src_default;
